kvdb/etcd/stm.go

```diff
@@ -320,42 +345,224 @@ func runSTM(s *stm, apply func(STM) error) error {
+
+// hasFullRange checks if the read set has a full range prefetched.
+func (rs *readSet) hasFullRange(prefix string) bool {
+	return rs.fullRanges[prefix]
+}
+
+// next returns the pre-fetched next value of the prefix. If matchKey is true,
+// it'll simply return the key/value that matches the passed key.
+func (rs *readSet) next(prefix, key string, matchKey bool) (*stmGet, bool) {
+	pivot := &stmGet{
+		KV: KV{
+			key: key,
+		},
+	}
+
+	var result *stmGet
+	rs.tree.AscendGreaterOrEqual(
+		pivot,
+		func(item btree.Item) bool {
+			next := item.(*stmGet)
+			if !matchKey && next.key == key {
+				return true
+			}
+
+			if next.rev != 0 && strings.HasPrefix(next.key, prefix) {
+				result = next
+			}
+
+			return false
+		},
+	)
+
+	return result, result != nil
+}
+
+// prev returns the pre-fetched prev key/value of the prefix from key.
+func (rs *readSet) prev(prefix, key string) (*stmGet, bool) {
+	pivot := &stmGet{
+		KV: KV{
+			key: key,
+		},
+	}
+
+	var result *stmGet
+
+	rs.tree.DescendLessOrEqual(
+		pivot, func(item btree.Item) bool {
+			prev := item.(*stmGet)
+			if prev.key == key {
+				return true
 			}
+
+			if prev.rev != 0 && strings.HasPrefix(prev.key, prefix) {
+				result = prev
+			}
+
+			return false
+		},
+	)
+
+	return result, result != nil
+}
+
+// last returns the last key/value of the passed range (if prefetched).
+func (rs *readSet) last(prefix string) (*stmGet, bool) {
+	key := []byte(prefix)
+	key[len(key)-1] += 1
+	return rs.prev(prefix, string(key))
+}
+
+// clear completely clears the readset.
+func (rs *readSet) clear() {
+	rs.tree.Clear(false)
+	rs.fullRanges = make(map[string]bool)
+}
+
+// getItem reurns the matching key/value from the readset.
+func (rs *readSet) getItem(key string) (*stmGet, bool) {
+	pivot := &stmGet{
+		KV: KV{
+			key: key,
+		},
+		rev: 0,
+	}
+	item := rs.tree.Get(pivot)
+	if item != nil {
+		return item.(*stmGet), true
+	}
+
+	// It's possible that although this key isn't in the read set, we
+	// fetched a full range the key is prefixed with. In this case we'll
+	// insert the key with zero revision.
+	for prefix := range rs.fullRanges {
+		if strings.HasPrefix(key, prefix) {
+			rs.tree.ReplaceOrInsert(pivot)
+			return pivot, true
 		}
 	}
+
+	return nil, false
 }
 
-// gets is a helper to create an op slice for transaction
-// construction.
-func (rs readSet) gets() []v3.Op {
-	ops := make([]v3.Op, 0, len(rs))
+// prefetchSet is a helper to create an op slice of all OpGet's that represent
+// fetched keys appended with a slice of all OpGet's representing all prefetched
+// full ranges.
+func (rs *readSet) prefetchSet() []v3.Op {
+	ops := make([]v3.Op, 0, rs.tree.Len())
+
+	rs.tree.Ascend(func(item btree.Item) bool {
+		key := item.(*stmGet).key
+		add := true
+		for prefix := range rs.fullRanges {
+			// Only add those items that have not been in prefetched
+			// full ranges to avoid unfetched when prefetching next.
+			if strings.HasPrefix(key, prefix) {
+				add = false
```

Comment:
Alternatively can just return `true` here early?

Reply:
Also if there're no prefixes at all `len(fullRanges) == 0`, then I think we can just return `false` here.